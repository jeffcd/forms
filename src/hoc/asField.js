import React from 'react'
import get from 'lodash/get'
import FormContext from '../FormContext'
import VisibilityContext from '../VisibilityContext'
import ListContext from '../ListContext'
import validateField, { validatorFunctions } from '../validateField'
import messages from '../messages'
const validators = validatorFunctions

const getValidators = props => {
  const fieldValidators = {}
  Object.keys(props).forEach(name => {
    if (validators[name]) {
      fieldValidators[name] = props[name]
    }
  })
  return fieldValidators
}

const errorMessages = messages

const getErrorMessages = ({ props, field }) => {
  return field.errors.map(error => {
    let message = errorMessages[error]
    if (props.messages && props.messages[error]) {
      message = props.messages[error]
    }
    return message
      .replace('%name%', props.label)
      .replace('%param%', props[error])
  })
}

const getInitialValue = (Field, { minLength = 0 }) => {
  if (Field.type === 'list') {
    const size = parseInt(minLength)
    return Array(size)
  }

  return ''
}

const asFieldHoc = Field => {
  return function asFieldHoc({ ...rest }) {
    return (
      <ListContext.Consumer>
        {listInfo => {
          const { name, initialValue, convertTo } = rest
          const fullName = listInfo
            ? `${listInfo.name}.value[${listInfo.i}].${name}`
            : name

          const listProps = {}
          if (listInfo) {
            const { i, name } = listInfo
            const fullName = `${name}.value[${i}].${rest.name}`
            const id = `${name}:${i}::${rest.name}:`
            listProps.id = id
            if (rest.name) {
              listProps.name = fullName
            }
          }

          return (
            <FormContext.Consumer>
              {form => {
                //useEffect(() => {
                //}, [form.setFormField])

                const field = get(form.fields, fullName)
                if (!field) {
                  const field = {
                    name,
                    value: initialValue || getInitialValue(Field, rest),
                    validators: getValidators(rest),
                    convertTo,
                    errors: []
                  }
                  const isUser = false
                  form.setFormField(field, fullName, isUser)

                  return null
                }

                return (
                  <VisibilityContext.Consumer>
                    {visibility => {
                      if (visibility.isVisible === false) {
                        if (field.isVisible !== false) {
                          const isUser = false
                          field.isVisible = false
                          form.setFormField({ ...field }, isUser)
                        }
                        return null
                      }

                      const handleChange = e => {
                        const field = get(form.fields, fullName)
                        const newField = {
                          ...field,
                          value: e.target.value
                        }
                        const errors = validateField(newField)
                        newField.errors = errors
                        form.setFormField(newField, fullName)
                      }
                      return (
                        <Field
                          onChange={handleChange}
                          value={field.value}
                          errors={getErrorMessages({ props: rest, field })}
                          {...rest}
                          {...listProps}
                        />
                      )
                    }}
                  </VisibilityContext.Consumer>
                )
              }}
            </FormContext.Consumer>
          )
        }}
      </ListContext.Consumer>
    )
  }
}

export default asFieldHoc
