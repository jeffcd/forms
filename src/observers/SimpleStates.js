import React from 'react'
import Typography from '@material-ui/core/Typography'
import { makeStyles } from '@material-ui/core/styles'
import withFormStateHoc from '../withFormStateHoc'

const stateStrs = {
  submit: 'Saving...',
  error: 'There are errors...',
  dirty: 'Unsaved changes...',
  pristine: ''
}

const useStyles = makeStyles(theme => ({
  SimpleStates: {
    marginTop: '16px',
    paddingLeft: theme.spacing(1)
  }
}))

const SimpleStates = ({ form }) => {
  const classes = useStyles()
  return (
    <Typography variant="body1" className={classes.SimpleStates}>
      {stateStrs[form.state]}
    </Typography>
  )
}

export default withFormStateHoc(SimpleStates)
