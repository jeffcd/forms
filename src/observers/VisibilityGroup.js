import React, { useState } from 'react'
import VisibilityContext from '../VisibilityContext'
import withFormStateHoc from '../hoc/withFormState'

const hidden = children => <div style={{ display: 'none' }}>{children}</div>

const SimpleStates = ({ form, isVisible, children }) => {
  const [visible, setVisible] = useState({})

  const inboundIsVisible = isVisible(form)
  if (visible.isVisible !== inboundIsVisible) {
    setVisible({ isVisible: inboundIsVisible })
  }

  return (
    <VisibilityContext.Provider value={visible}>
      {visible.isVisible !== false && children}
      {visible.isVisible === false && hidden(children)}
    </VisibilityContext.Provider>
  )
}

export default withFormStateHoc(SimpleStates)
