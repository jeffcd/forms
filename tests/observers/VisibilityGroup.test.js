import React from 'react'
import Form from '../../src'
import VisibilityGroup from '../../src/observers/VisibilityGroup'
import asField from '../../src/hoc/asField'
import SimpleStates from '../../src/observers/SimpleStates'

console.log(global.mount, mount)

const Input = asField(
  ({
    onChange,
    onBlur = () => {},
    name,
    id,
    value,
    errors,
    label,
    helperText,
    required = false
  }) => {
    return (
      <div>
        <label htmlFor={name}>{label}</label>
        <br />
        <input
          id={`input-${id}`}
          name={name}
          value={value}
          onChange={e => onChange(e)}
          onBlur={e => onBlur(e)}
          label={label}
          required={required}
        />
        <span>{errors.length ? errors[0] : helperText}</span>
      </div>
    )
  }
)

it('Form: Renders Empty', () => {
  const handleSubmit = jest.fn()
  const form = (
    <Form
      onSubmit={(e, form, actions) => {
        console.log(form)
        handleSubmit(e, form, actions)
      }}
    >
      <div>
        <Input name="age" label="Your Age" min="16" max="65" />
      </div>

      <VisibilityGroup
        isVisible={form => form.fields.age && form.fields.age.value > 18}
      >
        <Input name="numberJobs" label="Number of jobs..." />
        <VisibilityGroup
          isVisible={form =>
            form.fields.numberJobs && form.fields.numberJobs.value > 5
          }
        >
          Wow that is a lot!
          <Input
            name="managerRating"
            label="What would the average manager give on you a scale 1-10?"
          />
        </VisibilityGroup>
      </VisibilityGroup>
      <SimpleStates />
      <button type="submit" id="submit-button">
        Submit
      </button>
    </Form>
  )
  const tree = mount(form)

  expect(tree.html()).toMatchSnapshot()
  tree.find('form').simulate('submit')
  tree.find('#submit-button').simulate('click')
  expect(tree.html()).toMatchSnapshot()
  expect(handleSubmit).toHaveBeenCalledTimes(0)
})
