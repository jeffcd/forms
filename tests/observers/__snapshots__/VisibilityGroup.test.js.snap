// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form: Renders Empty 1`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><div><label for=\\"age\\">Your Age</label><br><input id=\\"input-undefined\\" name=\\"age\\" label=\\"Your Age\\" value=\\"\\"><span></span></div></div><div style=\\"display: none;\\"><div style=\\"display: none;\\">Wow that is a lot!</div></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Renders Empty 2`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><div><label for=\\"age\\">Your Age</label><br><input id=\\"input-undefined\\" name=\\"age\\" label=\\"Your Age\\" value=\\"\\"><span>Your Age must be greater or equal to 16.</span></div></div><div style=\\"display: none;\\"><div style=\\"display: none;\\">Wow that is a lot!</div></div>There are errors...<button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;
